src/utils_general/raster_manipulation.py

```diff
@@ -99,111 +101,90 @@ def fix_calendar(ds, timevar="F"):
-            nodata=nodata,
-        )
-    )["mean"]
-    df["avg_cell_touched"] = pd.DataFrame(
-        zonal_stats(
-            vectors=df,
-            raster=raster_array,
-            affine=raster_transform,
-            all_touched=True,
-            band=band,
-            nodata=nodata,
-        )
-    )["mean"]
-
-    # calculate the percentage of the area within an geographical area
-    # that has a value larger than threshold
-    forecast_binary = np.where(raster_array >= threshold, 1, 0)
-    bin_zonal = pd.DataFrame(
-        zonal_stats(
-            vectors=df,
-            raster=forecast_binary,
-            affine=raster_transform,
-            stats=["count", "sum"],
-            band=band,
-            nodata=nodata,
-        )
-    )
-    df["perc_threshold"] = bin_zonal["sum"] / bin_zonal["count"] * 100
-    bin_zonal_touched = pd.DataFrame(
-        zonal_stats(
-            vectors=df,
-            raster=forecast_binary,
-            affine=raster_transform,
-            all_touched=True,
-            stats=["count", "sum"],
-            band=band,
-            nodata=nodata,
-        )
-    )
-    df["perc_threshold_touched"] = (
-        bin_zonal_touched["sum"] / bin_zonal_touched["count"] * 100
-    )
-
-    return df
+    df_list = []
+
+    if stats_list is None:
+        stats_list = ["mean", "std", "min", "max", "sum", "count"]
+
+    for bound_id in gdf[bound_col].unique():
+        gdf_adm = gdf[gdf[bound_col] == bound_id]
+
+        da_clip = raster_array.rio.set_spatial_dims(
+            x_dim=lon_coord, y_dim=lat_coord
+        ).rio.clip(gdf_adm[geom_col], all_touched=all_touched)
+
+        grid_stat_all = []
+        for stat in stats_list:
+            # count automatically ignores NaNs
+            # therefore skipna can also not be given as an argument
+            # implemented count cause needed for computing percentages
+            kwargs = {} if stat == "count" else {"skipna": True}
+            # makes sum return NaN instead of 0 if array
+            # only contains NaNs
+            if stat == "sum":
+                kwargs["min_count"] = 1
+            grid_stat = getattr(da_clip, stat)(
+                dim=[lon_coord, lat_coord], **kwargs
+            )
+            grid_stat = grid_stat.rename(f"{stat}_{bound_col}")
+            grid_stat_all.append(grid_stat)
+        if percentile_list is not None:
+            for quant in percentile_list:
+                quant_float = quant / 100
+                grid_quant = da_clip.quantile(
+                    quant_float, dim=[lon_coord, lat_coord]
+                )
+                grid_quant = grid_quant.drop("quantile").rename(
+                    f"{quant}quant_{bound_col}"
+                )
+                grid_stat_all.append(grid_quant)
+
+        # if dims is 0, it throws an error when merging
+        # and then converting to a df
+        # this occurs when the input da is 2D
+        if grid_stat_all[0].dims == ():
```

Comment:
This could be made slightly more general by writing `if not grid_stat_all[0].dims` (that way if dims is a different type but also empty it will still work)